[WIP] Stylesheets as part of documentation code base

stejskalleos opened the pull request:
WIP PR for stylesheet management as part of the documentation code base

* No need for https://github.com/lzap/asciidoctor-stylesheet-factory repo anymore
* Easier editing styles

**TODO**
* Update README.md
* Github actions
* Cherry-picks?

**Notes**
Code & logic copy-pasted from Lzap's https://github.com/lzap/asciidoctor-stylesheet-factory repo repository without any changes, except the cleanup of unused `sass` files

??? Cherry-pick into:

* [ ] Foreman 3.0
* [ ] Foreman 2.5 (Satellite 6.10)
* [ ] Foreman 2.4
* [ ] Foreman 2.3 (Satellite 6.9)
* [ ] Foreman 2.1 (Satellite 6.8)

<!---
Thank you for contributing to Foreman documentation. Make sure to read README
for the documentation standards. Set cherry-pick github label to mark this
contribution for cherry picking and check which version do you need with [x].
-->

lzap commented:
Ok I see no changes of generated HTML guides, that's what's expected, is it?

lzap commented:
Tests do fail due to broken HTML links, but it compiles fine. Any comments @ekohl ? LGTM

lzap commented:
Ah tiny problem, add the following into `.gitignore` please:

```
 M guides/common/default.css
?? .bundle/
?? Gemfile.lock
```

lzap commented:
Please rebase on top of `master`, commit https://github.com/theforeman/foreman-documentation/commit/af69eddb2908dc3778841ae0ca89c4747176db58 fixed HTML validation errors on CI.

stejskalleos commented:
Rebased & updated

> Ok I see no changes of generated HTML guides, that's what's expected, is it?

Correct, no changes in the UI

maximiliankolb commented on guides/README.md:
```diff
@@ -11,19 +11,19 @@ Contributions are welcome. Please read the [Contribution guidelines](#contributi
 		xcode-select --install
 
-If AsciiDoctor is not available in repositories or under RVM/rbenv, simply install it from rubygems:
+Install ruby gems (in documentation folder):
```
Do you mean in `foreman-documentation` directory?
```suggestion
Install ruby gems in `foreman-documentation` folder:
```

maximiliankolb commented on guides/common/default.css:
```diff
@@ -1,780 +1 @@
-		box-shadow: 4px 4px 10px rgba(0, 0, 0, 0.8);
-	}
-}
+@import url(http://red-hat-overpass-fonts.s3.amazonaws.com/overpass.css);/*! normalize.css v2.1.2 | MIT License | git.io/normalize */article,aside,details,figcaption,figure,footer,header,hgroup,main,nav,section,summary{display:block}audio,canvas,video{display:inline-block}audio:not([controls]){display:none;height:0}[hidden],template{display:none}script{display:none !important}html{font-family:san ...
```
I don't like this change at all. It makes it very hard to review and view git diffs. Can you please format the `.css` file in a human friendly way?

maximiliankolb commented on guides/common/sass/components/_asciidoc.scss:
```diff
@@ -0,0 +1,1489 @@
+// QUESTION: should we add max-width:100% and height:auto to svg?
```
Do we have to address this question in this PR? Do you want to create an issue for that?

stejskalleos replied:
This is copy/pasted file from asci-doc styles, not my comment

maximiliankolb replied:
OK, then disregard my questions.

maximiliankolb commented on guides/common/sass/components/_asciidoc.scss:
```diff
@@ -0,0 +1,1489 @@
+  font-size: $paragraph-font-size + em-calc(3.5);
+  line-height: 1.6;
+  color: $title-font-color;
+  // QUESTION remove or reduce letter-spacing?
```
See above.

stejskalleos replied:
same as above, not mine code, it's from ascii-doc team

maximiliankolb commented on guides/common/sass/settings/_defaults.scss:
```diff
@@ -0,0 +1,101 @@
+@import "foundation/components/global";
+@import "compass/css3/box-shadow";
+
+// better defaults (TODO maybe do these at the bottom of the theme colors)
```
action required? or keep as is for now?

stejskalleos replied:
same as above, not mine code, it's from ascii-doc team

maximiliankolb reviewed (changes requested):
I did not test/check each line of css, but looks reasonable overall.

I have a strong option on the css formatting. Apart from that, there are a couple of `todo`, `question`, and `fixme` comments. Do we need to address them in this PR or open an issue?

stejskalleos commented:
Rebased & updated

> I have a strong option on the css formatting.

There was a bug while compiling the assets, now fixed

> Apart from that, there are a couple of todo, question, and fixme comments. Do we need to address them in this PR or open an issue?

As per comments, this is from asciidoc team, not us.

lzap commented on guides/common/Makefile:
```diff
@@ -48,6 +50,7 @@ open-pdf: pdf
 
 $(DEST_DIR)/$(BUILD).css: $(COMMON_DIR)/$(BUILD).css
+	bundle exec compass compile --force -c $(COMMON_DIR)/sass_config.rb --sass-dir $(COMMON_DIR)/sass --css-dir $(COMMON_DIR);
 	cp $< $@
```
This is not correct, in Makefile the syntax is:

```
target_file: sourcefile1 sourcefile2 ...etc
  command to run
```

Previously the css target was a simple copy via `cp`. You want to change that to run a different command (`compass` in this case). You want it to create `$(DEST_DIR)/$(BUILD).css`, I am not sure how to do this tho, the utility do not appear to have any `--output` argument to specify where to store the result. Do not store anything in the common directory, use the build directory for the output.

It is very important that the target file and source(s) are valid paths, Makefile is basically a dependency solving tool. What you have here does not work.

lzap commented:
Please update it accordingly to get Makefile working as expected:

```diff
diff --git a/guides/common/Makefile b/guides/common/Makefile
index c6719d30..4f55fa7d 100644
--- a/guides/common/Makefile
+++ b/guides/common/Makefile
@@ -9,6 +9,7 @@ DEST_HTML = $(DEST_DIR)/index-$(BUILD).html
 DEST_PDF = $(BUILD_DIR)/$(subst doc-,,$(NAME))-$(BUILD).pdf
 IMAGES_DIR = $(DEST_DIR)/images
 IMAGES_TS = $(DEST_DIR)/.timestamp-images
+CSS_SOURCES = $(shell find $(COMMON_DIR)/sass -type f)
 SOURCES = master.adoc $(shell find . -type f -name \*.adoc)
 DEPENDENCIES = $(shell find $(COMMON_DIR) -type f -name \*.adoc)
 IMAGES = $(shell find ./images ./common/images -type f 2>/dev/null || true)
@@ -33,8 +34,6 @@ prepare:
        mkdir -p $(BUILD_DIR) $(DEST_DIR) $(IMAGES_DIR) $(IMAGES_DIR) $(DEST_DIR)/common/images

 clean:
-       @rm -rf $(COMMON_DIR)/default.css
-       @touch $(COMMON_DIR)/default.css
        @rm -rf "$(DEST_DIR)" "$(DEST_PDF)"

 browser: html
@@ -49,8 +48,10 @@ linkchecker: html
 open-pdf: pdf
        ${BROWSER_OPEN} "$(realpath $(ROOTDIR)/$(DEST_PDF))"

-$(DEST_DIR)/$(BUILD).css: $(COMMON_DIR)/$(BUILD).css
-       bundle exec compass compile --force -c $(COMMON_DIR)/sass_config.rb --sass-dir $(COMMON_DIR)/sass --css-dir $(COMMON_DIR);
+$(BUILD_DIR)/default.css: $(CSS_SOURCES)
+       bundle exec compass compile --force -c $(COMMON_DIR)/sass_config.rb --sass-dir $(COMMON_DIR)/sass --css-dir $(BUILD_DIR)
+
+$(DEST_DIR)/$(BUILD).css: $(BUILD_DIR)/default.css
        cp $< $@

 $(IMAGES_TS): $(IMAGES)
```

With this change it will work as it should:

* make clean - cleans build directory
* make html - builds default.css first, then copy it to the destination under proper name
* when any SASS file or RB config is changed, CSS is rebuilt
* slower compass is only executed once (then the result is copied to the destinations via `cp`)
*

stejskalleos commented:
> Are you going to continue improving CSS now? What was the motivation again?

Correct, my plan is to improve some styles & add new stuff
